tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,376 @@
+  private readonly rgbColors = ['240, 120, 80', '46, 119, 182', '190, 64, 36'];
+  private readonly area = d3
+    .area<ViolinBin>()
+    .x0(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(-d.length);
+      }.bind(this)
+    )
+    .x1(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(d.length);
+      }.bind(this)
+    )
+    .y(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        if (d.x0! === -Infinity) {
+          return this.chartHeight - this.drawMargin.top;
+        }
+        return this.yScale((d.x1! + d.x0!) / 2.0);
+      }.bind(this)
+    )
+    .curve(d3.curveCatmullRom);
+
+  ngAfterViewInit(): void {
+    this.svg = d3.select(this.chartContainer.nativeElement).select('svg');
+    this.svg;
+    this.mainContainer = this.svg
+      .append('g')
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.drawContainer = this.mainContainer
+      .append('g')
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left}, ${this.drawMargin.top})`
+      );
+    this.dotsGroup = this.drawContainer.append('g').attr('class', 'dotsGroup');
+    this.yAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--y');
+    this.xAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--x');
+    this.miscGroup = this.drawContainer.append('g');
+    this.xScale = d3.scaleBand().padding(0.05);
+    this.yScale = d3.scaleLinear().range([this.drawHeight, 0]);
+    this.xScaleNum = d3.scaleLinear();
+    this.initializeBrush();
+    this.drawMisc();
+    this.redraw();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.svg) {
+      this.redraw();
+    }
+  }
+
+  redraw() {
+    this.setMaxBinSize();
+    this.updateAxes();
+    this.draw();
+  }
+
+  // Initializing/Updating the visualization props
+  setMaxBinSize() {
+    Object.values(this.chartData.violinData).forEach((dataElement) => {
+      const lengths = dataElement.map((bin) => bin.length);
+      const longest: number = Math.max(...lengths);
+      this.maxBinSize = Math.max(longest, this.maxBinSize);
+    });
+  }
+
+  updateAxes() {
+    this.xScale
+      .range([0, this.drawWidth])
+      .domain(Object.keys(this.chartData.violinData));
+    this.xAxis = d3.axisBottom(this.xScale);
+
+    this.yScale.domain([
+      this.chartData.extremes.min,
+      this.chartData.extremes.max,
+    ]);
+    this.yAxis = d3.axisLeft(this.yScale);
+
+    this.xScaleNum
+      .range([0, this.xScale.bandwidth()])
+      .domain([-this.maxBinSize, this.maxBinSize]);
+  }
+
+  initializeBrush() {
+    this.brush.on('end', this.brushMoved.bind(this));
+  }
+
+  // Drawing UI
+  draw() {
+    this.drawAxes();
+    this.drawPlot();
+    this.refreshMisc();
+    this.refreshBrush();
+  }
+
+  drawAxes() {
+    this.miscGroup.selectAll('*').remove();
+    this.miscGroup
```

Comment:
Minor but I am a bit concerned about this: 

because drawAxes remove everything under miscGroup and miscGroup is both modified by drawAxes and drawMisc, they need to always go hand in hand. (1) can you not remove all the elements on redraw or (2) can you factor out the DOM removal at the top of the `draw` or `redraw` method?

Reply:
Forgot this. Thanks for the pointer.